button-listen.c

```diff
@@ -11,37 +14,77 @@ static double timespec_diff(struct timespec const* start, struct timespec const*
     return s + (ns / 1e9);
 }
 
+static bool open_file(FILE **evf, const char *filename)
+{
+    FILE *file = NULL;
+    int fd = -1;
+    file = fopen(filename, "rb");
+    if(!file) {
+        fprintf(stderr, "Could not open %s\n", filename);
+        return false;
+    }
+    fd = fileno(file);
+    if (fd < 0 || fd >= FD_SETSIZE) {
+        fprintf(stderr, "File descriptor for %s too large to handle (%d > %d)", filename, fd, FD_SETSIZE);
+        return false;
+    }
+    evf[fd] = file;
+    return true;
+}
+
+void prep_evf_set(FILE **evf, fd_set* fd_set) {
+    for (int i = 0; i < FD_SETSIZE; i++){
+        if(evf[i]) {
+            FD_SET(i, fd_set);
+        }
+    }
+}
+
 int main(int argc, char** argv)
 {
-    FILE* evf = fopen("/dev/input/event2", "rb");
-    if(!evf) {
-        fprintf(stderr, "Could not open /dev/input/event2\n");
+    FILE *evf[FD_SETSIZE];
+    for (int i =0; i < FD_SETSIZE; i++){
+        evf[i] = NULL;
+    }
+    if(!open_file(evf, "/dev/input/event2")) {
         return 1;
     }
-
+    open_file(evf, "/dev/input/event1");
     struct input_event ev;
     struct timespec press_time = {0};
     bool pressed = false;
     while(1) {
-        size_t sz = fread(&ev, sizeof(ev), 1, evf);
-        if(sz == 0) {
+        fd_set evf_set;
+        FD_ZERO(&evf_set);
+        prep_evf_set(evf, &evf_set);
+        if(select(FD_SETSIZE, &evf_set, NULL, NULL, NULL) < 0){
```

Comment:
I'd prefer an actual computed `nfds` instead of just throwing `FD_SETSIZE` at `select`, but that's probably minor in the grand scheme of things ;).

(That, or just ditching select's clunky API and moving to poll ;p).

Reply:
Good point. I'd not used `poll` before but that is **way** nicer.